### PR TITLE
Add video support to ffmpeg encoder/decoder

### PR DESCRIPTION
**Pull request by toots**
This PR adds support for video encoding and decoding using ffmpeg's native library!

It is ready to be tested! Here's a local test one-liner I've been using:
```
output.file(%ffmpeg(video_codec="libx264",format="avi"),"/tmp/test.avi",single("/tmp/bla.mp4"))
```

giphy

### CODE REVIEW

**Comment by smimram** [comment]
Oh yeah!!!

File: src/encoder.ml [hunk 1]
```diff
@@ -51,8 +51,17 @@ let kind_of_format = function
...
   | Ffmpeg m ->
-      { Frame.audio = m.Ffmpeg_format.channels ;
-        Frame.video = 0 ; Frame.midi = 0 }
+      let channels = 
+        match m.Ffmpeg_format.audio_codec with
+          | None -> 0
+          | Some _ -> m.Ffmpeg_format.channels
+      in
+      let video =
+        if m.Ffmpeg_format.video_codec <> None then 1 else 0
```
Comment (smimram) [inline src/encoder.ml]: Is there a slight possibility that we could support formats with multiple video channels?

File: src/stream/generator.ml [hunk 2]
```diff
@@ -114,19 +114,19 @@ let remove_end g remove_len =
    * possibly cut some element in half, remove the rest.
    * The parsed elements are put back at the end
    * of the queue. *)
-  let rec cut l =
+  let rec cut ?(initial_ofs=0) l =
     let (c,ofs,len) = Queue.take g.buffers in
-      if len<l then begin
+      if len-initial_ofs<l then begin
         Queue.push (c,ofs,len) g.buffers ;
```
Comment (smimram) [inline src/stream/generator.ml]: I guess we could improve things by having buffers with view as for `Strings` here... (but this is not really related to the current PR).

File: src/encoder.ml (hunk 1 above)
Comment (toots) [reply src/encoder.ml]: It is supported in theory, yes.

File: src/stream/generator.ml (hunk 2 above)
Comment (toots) [reply src/stream/generator.ml]: Agreed. The implementation of these buffers is not straight forward and could be enhanced indeed.

**Comment by toots** [comment]
Youtube streaming works now!

```
s = single("/tmp/bla.mp4")

encoder = %ffmpeg(video_codec="libx264",audio_codec="libmp3lame",channels=1,
                  format="flv",url="rtmp://x.rtmp.youtube.com/live2/<key>")

output.null(encoder,s)
```

**Comment by Infernion** [comment]
@toots is it possible to decode rtmp stream in this way?

**Comment by toots** [comment]
> @toots is it possible to decode rtmp stream in this way? Also, could you point me how could I install version from this branch? I've found this [one](https://github.com/savonet/liquidsoap/blob/master/doc/content/install.md#latest-development-version) but seems it doesn't work on my machine

The problem for inputs is that we need a way to receive the stream, which is not included in the ffmpeg library API -- at least as far as I know.. I'll look into it. Natively, we only support HTTP inputs via the harbor for now and some experimental UDP input.

Re: installing and testing the branch, you can either use the CI-based debian packages and install `liquidsoap-ffmpeg-video` or use opam with pinning. In this case, you will need the latest `ocaml-ffmpeg`:
```
git clone https://github.com/savonet/ocaml-ffmpeg.git
cd ocaml-ffmpeg
opam pin add .

git clone https://github.com/savonet/liquidsoap.git
cd liquidsoap && git checkout ffmpeg-video
opam pin add .
```

**Comment by toots** [comment]
> > @toots is it possible to decode rtmp stream in this way? Also, could you point me how could I install version from this branch? I've found this [one](https://github.com/savonet/liquidsoap/blob/master/doc/content/install.md#latest-development-version) but seems it doesn't work on my machine
> 
> The problem for inputs is that we need a way to receive the stream, which is not included in the ffmpeg library API -- at least as far as I know.. I'll look into it. Natively, we only support HTTP inputs via the harbor for now and some experimental UDP input.

Ok, good news, it seems that this shouldn't be too hard to implement. Stay tuned!

**Comment by toots** [comment]
@Infernion Just added `input.ffmpeg`!

**Comment by Infernion** [comment]
@toots I couldn't run it. I've downloaded last version with apt install. 
This is installed version:

```
$  liquidsoap --version                                                        
Liquidsoap 1.5.0+git@869e278a                                                                                          
Copyright (c) 2003-2019 Savonet team                                                                                   
Liquidsoap is open-source software, released under GNU General Public License.                                         
See <http://liquidsoap.info> for more information.
```

When I try run script [above](https://github.com/savonet/liquidsoap/pull/1038#issuecomment-559128035) I've gotten: `Error 4: Undefined variable output.url`

**Comment by dhannyz** [comment]
@toots could input.ffmpeg also be adapted to run listener type inputs to run as an ingest server?
Eg of what im using with input.external `ffmpeg -re -listen 1 -i rtmp://IP:PORT/MOUNT ...`

**Comment by toots** [comment]
@Infernion did you recompile/reinstall with the latest `ocaml-ffmpeg` and `liquidsoap` `master` branch?

@dhannyz I'll look into it, it shouldn't be too hard to support I believe.

**Comment by toots** [comment]
One more thing: @dhannyz a separate authentication layer using ffmpeg's server input capabilities will most likely not be possible, I fear, unless they provide an API for it, which I haven't seen yet.

**Comment by dhannyz** [comment]
@toots it's unfortunate, and will add a bit of latency, but i can handle auth external to liquidsoap, something like nginx-rtmp forwarding to liquidsoap locally.
I did have a quick read through libav's rtmpproto.c and there doesnt seem to be any auth handling in the listen mode. The most it does is log a warning when the 'app' doesnt match that set in the listener setup

**Comment by Infernion** [comment]
I've tried to fresh install latest commits and the following error has been appeared:

```
...
OCAMLOPT -c decoder/ffmpeg_decoder.ml
File "decoder/ffmpeg_decoder.ml", line 121, characters 31-51:
121 |     match List.find_opt (fun ({FFmpeg.Avfilter.name}) ->
                                     ^^^^^^^^^^^^^^^^^^^^
Error: Unbound record field FFmpeg.Avfilter.name
make[2]: *** [decoder/ffmpeg_decoder.cmx] Error 2
make[1]: *** [all-auto-ocaml-prog] Error 2
make: *** [all-subdirs] Error 2
```

**Comment by smimram** [comment]
You need to update ocaml-ffmpeg too (git pull in the corresponding directory).

**Comment by Infernion** [comment]
@smimram I have latest versions. The last commit of ocaml-ffmpeg: 2dc8ad04d6a9e6935e5435c93d21cac9f40dd367

```
➜  ocaml-ffmpeg git:(master) opam pin add .
[NOTE] Package ffmpeg is already pinned to
       git+file:///Users/serhii/Projects/ocaml-ffmpeg#master (version 0.4.2).
[ffmpeg.0.4.2] synchronised from git+file:///Users/serhii/Projects/ocaml-ffmpeg#master
ffmpeg is now pinned to git+file:///Users/serhii/Projects/ocaml-ffmpeg#master (version 0.4.2)
Already up-to-date.
Nothing to do.
```

And the error is same :) Maybe I should wait when 1.5.0 would be released, but I have desire for testing XD

**Comment by smimram** [comment]
I can confirm that this is the lastest commit for ffmpeg, but I can also confirm that Liquidsoap is building alright here:) Are you sure you don't have another copy of the ffmpeg library (like a package or so).

**Comment by Infernion** [comment]
@smimram I confirm that I have a little bit mess with dependencies under opam. But it's actually hard to install all needed dependencies when they're stored only in git. It requires manual cloning. I've seen the liquidsoap-full repo where stored all dependencies but it was hard to me manage with them. I do not actually know with dependencies required

**Comment by toots** [comment]
> @smimram I confirm that I have a little bit mess with dependencies under opam. But it's actually hard to install all needed dependencies when they're stored only in git. It requires manual cloning. I've seen the liquidsoap-full repo where stored all dependencies but it was hard to me manage with them. I do not actually know with dependencies required

`opam` can be a little annoying when dealing with `git` sources. My workflow is to call `opam install -y .` from the repo to force the install of the current commit.

**Comment by Infernion** [comment]
Ok, I've figured out how to install. Now it's kind work. I've run code from [comment](https://github.com/savonet/liquidsoap/pull/1038#issuecomment-559128035) and get segment fault.

```
liquidsoap --debug ser.liq           
2019/12/13 20:31:15 >>> LOG START
2019/12/13 20:31:15 [main:3] Liquidsoap 1.5.0+git@ab4332bd
2019/12/13 20:31:15 [main:3] Using: bytes=[distributed with OCaml 4.02 or above] pcre=7.4.3 sedlex=2.1 menhirLib=20190924 dtools=0.4.1 duppy=0.8.1 mm=0.6.0 dynlink=[distributed with Ocaml] ffmpeg=0.4.2 camomile=1.0.2
2019/12/13 20:31:15 [main:3] 
2019/12/13 20:31:15 [main:3] DISCLAIMER: This version of Liquidsoap has been
2019/12/13 20:31:15 [main:3] compiled from a snapshot of the development code.
2019/12/13 20:31:15 [main:3] As such, it should not be used in production
2019/12/13 20:31:15 [main:3] unless you know what you are doing!
2019/12/13 20:31:15 [main:3] 
2019/12/13 20:31:15 [main:3] We are, however, very interested in any feedback
2019/12/13 20:31:15 [main:3] about our development code and committed to fix
2019/12/13 20:31:15 [main:3] issues as soon as possible.
2019/12/13 20:31:15 [main:3] 
2019/12/13 20:31:15 [main:3] If you are interested in collaborating to
2019/12/13 20:31:15 [main:3] the development of Liquidsoap, feel free to
2019/12/13 20:31:15 [main:3] drop us a mail at <savonet-devl@lists.sf.net>
2019/12/13 20:31:15 [main:3] or to join the slack chat at <http://slack.liquidsoap.info>.
2019/12/13 20:31:15 [main:3] 
2019/12/13 20:31:15 [main:3] Please send any bug report or feature request
2019/12/13 20:31:15 [main:3] at <https://github.com/savonet/liquidsoap/issues>.
2019/12/13 20:31:15 [main:3] 
2019/12/13 20:31:15 [main:3] We hope you enjoy this snapshot build of Liquidsoap!
2019/12/13 20:31:15 [main:3] 
2019/12/13 20:31:15 [dynamic.loader:3] Could not find dynamic module for lame encoder.
2019/12/13 20:31:15 [dynamic.loader:3] Could not find dynamic module for fdkaac encoder.
2019/12/13 20:31:15 [decoder:4] Trying method "META" for "/Users/serhii/Downloads/2D73DDB8-33E9-4FF7-ACD4-FA7FB1296FC6.mp4"...
2019/12/13 20:31:15 [decoder:4] Trying method "WAV" for "/Users/serhii/Downloads/2D73DDB8-33E9-4FF7-ACD4-FA7FB1296FC6.mp4"...
2019/12/13 20:31:15 [decoder.wav/aiff:4] Invalid file extension for "/Users/serhii/Downloads/2D73DDB8-33E9-4FF7-ACD4-FA7FB1296FC6.mp4"!
2019/12/13 20:31:15 [decoder:4] Trying method "AIFF" for "/Users/serhii/Downloads/2D73DDB8-33E9-4FF7-ACD4-FA7FB1296FC6.mp4"...
2019/12/13 20:31:15 [decoder.wav/aiff:4] Invalid file extension for "/Users/serhii/Downloads/2D73DDB8-33E9-4FF7-ACD4-FA7FB1296FC6.mp4"!
2019/12/13 20:31:15 [decoder:4] Trying method "MIDI" for "/Users/serhii/Downloads/2D73DDB8-33E9-4FF7-ACD4-FA7FB1296FC6.mp4"...
2019/12/13 20:31:15 [decoder:4] Trying method "IMAGE" for "/Users/serhii/Downloads/2D73DDB8-33E9-4FF7-ACD4-FA7FB1296FC6.mp4"...
2019/12/13 20:31:15 [decoder:4] Trying method "FFMPEG" for "/Users/serhii/Downloads/2D73DDB8-33E9-4FF7-ACD4-FA7FB1296FC6.mp4"...
2019/12/13 20:31:15 [decoder.ffmpeg:4] ffmpeg recognizes "/Users/serhii/Downloads/2D73DDB8-33E9-4FF7-ACD4-FA7FB1296FC6.mp4" as: audio: {codec: aac, 44100Hz, 2 channel(s)}, video: {codec: h264, 1920x960, yuv420p}.
2019/12/13 20:31:15 [decoder:3] Method "FFMPEG" accepted "/Users/serhii/Downloads/2D73DDB8-33E9-4FF7-ACD4-FA7FB1296FC6.mp4".
2019/12/13 20:31:15 [decoder.id3v2:4] Invalid file extension for "/Users/serhii/Downloads/2D73DDB8-33E9-4FF7-ACD4-FA7FB1296FC6.mp4"!
2019/12/13 20:31:15 [frame:3] Using 44100Hz audio, 25Hz video, 44100Hz master.
2019/12/13 20:31:15 [frame:3] Frame size must be a multiple of 1764 ticks = 1764 audio samples = 1 video samples.
2019/12/13 20:31:15 [frame:3] Targetting 'frame.duration': 0.04s = 1764 audio samples = 1764 ticks.
2019/12/13 20:31:15 [frame:3] Frames last 0.04s = 1764 audio samples = 1 video samples = 1764 ticks.
2019/12/13 20:31:15 [sandbox:3] Could not find binary bwrap, disabling sandboxing..
2019/12/13 20:31:15 [video.converter:3] Couldn't find preferred video converter: gavl.
2019/12/13 20:31:15 [audio.converter:3] Using samplerate converter: ffmpeg.
2019/12/13 20:31:15 [threads:4] Created thread "generic queue #1" (1 total).
2019/12/13 20:31:15 [threads:4] Created thread "generic queue #2" (2 total).
2019/12/13 20:31:15 [threads:4] Created thread "non-blocking queue #1" (3 total).
2019/12/13 20:31:15 [threads:4] Created thread "non-blocking queue #2" (4 total).
2019/12/13 20:31:15 [clock:4] Currently 1 clocks allocated.
2019/12/13 20:31:15 [clock.main:4] Starting 1 sources...
2019/12/13 20:31:15 [source:4] Source output.file_7114 gets up.
2019/12/13 20:31:15 [source:4] Source single_7111 gets up.
2019/12/13 20:31:15 [single_7111:3] "/Users/serhii/Downloads/2D73DDB8-33E9-4FF7-ACD4-FA7FB1296FC6.mp4" is static, resolving once for all...
2019/12/13 20:31:15 [single_7111:4] Content kind is {audio=2;video=1;midi=0}.
2019/12/13 20:31:15 [single_7111:4] Activations changed: static=[output(dot)url:output(dot)url], dynamic=[].
2019/12/13 20:31:16 [single_7111:3] Prepared "/Users/serhii/Downloads/2D73DDB8-33E9-4FF7-ACD4-FA7FB1296FC6.mp4" (RID 0).
2019/12/13 20:31:16 [output(dot)url:4] Activations changed: static=[output(dot)url], dynamic=[].
2019/12/13 20:31:16 [output(dot)url:4] Enabling caching mode: active source.
2019/12/13 20:31:18 [threads:4] Created thread "clock_main" (1 total).
2019/12/13 20:31:18 [clock:4] Main phase starts.
2019/12/13 20:31:18 [clock.main:3] Streaming loop starts in auto-sync mode
2019/12/13 20:31:18 [clock.main:3] Delegating synchronisation to CPU clock
[2]    98163 segmentation fault  liquidsoap --debug ser.liq
```

I'm not sure how to debug it with gdb

**Comment by toots** [comment]
Thanks for taking the time to test! I just rebuilt with the latest version and it works here.

You can try this:
```
gdb /path/to/liquidsoap
> r <options>
.. crashes ..
> thread apply all bt
```

**Comment by smimram** [comment]
I prefer
```
gdb --args liquidsoap myscript.liq
> r
... [crash]
> thread apply all bt
```
and yes it would be really helpful if you could give us the output.

**Comment by Infernion** [comment]
I have the following output:

```
Reading symbols from liquidsoap...
(gdb) r
Starting program: /Users/serhii/.opam/4.09.0/bin/liquidsoap ser.liq
[New Thread 0x1003 of process 3385]
[New Thread 0xc03 of process 3385]
warning: unhandled dyld version (16)
init: security exit, root euid & guid (user & group). Override with set("init.allow_root",true)
[Inferior 1 (process 3385) exited with code 0377]
(gdb) thread apply all bt
(gdb)
```

**Comment by Infernion** [comment]
Here is the output:
```
...
2019/12/15 16:26:14 [single_7111:3] Prepared "/Users/serhii/Movies/348.mp4" (RID 0).
2019/12/15 16:26:15 [clock.main:3] Streaming loop starts in auto-sync mode
2019/12/15 16:26:15 [clock.main:3] Delegating synchronisation to CPU clock
[New Thread 0xf07 of process 3484]
[New Thread 0x1103 of process 3484]
[New Thread 0x1503 of process 3484]
[New Thread 0x1603 of process 3484]
[New Thread 0x1703 of process 3484]
[New Thread 0x1803 of process 3484]
[New Thread 0x1903 of process 3484]
[New Thread 0x1a03 of process 3484]
[New Thread 0x2303 of process 3484]
[New Thread 0x2403 of process 3484]
[New Thread 0x2503 of process 3484]
[New Thread 0x2603 of process 3484]
[New Thread 0x2703 of process 3484]
[New Thread 0x2803 of process 3484]
[New Thread 0x2903 of process 3484]

Thread 3 received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0xf07 of process 3484]
0x0000000100e78987 in ?? ()
(gdb) thread apply all bt

Thread 17 (Thread 0x2903 of process 3484):
#0  0x00007fff6970c896 in ?? ()
#1  0x0000000100352def in caml_poll (_read=4433695888, _write=4302152488, _err=4302152488, _timeout=<optimized out>) at duppy_stubs.c:95
#2  0x00000001002a5460 in camlDuppy__poll_124 ()
#3  0x00000001002a65f0 in camlDuppy__f_486 ()
#4  0x00000001002a61cb in camlDuppy__process_477 ()
#5  0x00000001002a6f41 in camlDuppy__f_563 ()
#6  0x00000001002a6ca6 in camlDuppy__queue_inner_2442 ()
#7  0x00000001001cc826 in camlTutils__queue_761 ()
#8  0x00000001001cbfb9 in camlTutils__fun_1467 ()
#9  0x00000001002d6e0b in camlThread__fun_286 ()
#10 0x0000000100382ed8 in caml_float_pcm_convert_s24le (_src=4383047936, _offset=3, _dst=4294967295) at audio_c.c:411
#11 0x0000000100357188 in caml_thread_start ()
#12 0x00007fff697c9e65 in ?? ()
#13 0x0000000000000000 in ?? ()

Thread 16 (Thread 0x2803 of process 3484):
#0  0x00007fff69708ce6 in ?? ()
#1  0x00007fff697ca185 in ?? ()
#2  0x0000000000000000 in ?? ()

--Type <RET> for more, q to quit, c to continue without paging--
Thread 15 (Thread 0x2703 of process 3484):
#0  0x00007fff69708ce6 in ?? ()
#1  0x00007fff697ca185 in ?? ()
#2  0x0000000000000000 in ?? ()

Thread 14 (Thread 0x2603 of process 3484):
#0  0x00007fff69708ce6 in ?? ()
#1  0x00007fff697ca185 in ?? ()
#2  0x0000000000000000 in ?? ()

Thread 13 (Thread 0x2503 of process 3484):
#0  0x00007fff69708ce6 in ?? ()
#1  0x00007fff697ca185 in ?? ()
#2  0x0000000000000000 in ?? ()

Thread 12 (Thread 0x2403 of process 3484):
#0  0x00007fff69708ce6 in ?? ()
#1  0x00007fff697ca185 in ?? ()
#2  0x0000000000000000 in ?? ()

Thread 11 (Thread 0x2303 of process 3484):
#0  0x00007fff69708ce6 in ?? ()
#1  0x00007fff697ca185 in ?? ()
#2  0x0000000000000000 in ?? ()
--Type <RET> for more, q to quit, c to continue without paging--

Thread 10 (Thread 0x1a03 of process 3484):
#0  0x00007fff69708ce6 in ?? ()
#1  0x00007fff697ca185 in ?? ()
#2  0x0000000000000000 in ?? ()

Thread 9 (Thread 0x1903 of process 3484):
#0  0x00007fff69708ce6 in ?? ()
#1  0x00007fff697ca185 in ?? ()
#2  0x0000000000000000 in ?? ()

Thread 8 (Thread 0x1803 of process 3484):
#0  0x00007fff69708ce6 in ?? ()
#1  0x00007fff697ca185 in ?? ()
#2  0x0000000000000000 in ?? ()

Thread 7 (Thread 0x1703 of process 3484):
#0  0x00007fff69708ce6 in ?? ()
#1  0x00007fff697ca185 in ?? ()
#2  0x0000000000000000 in ?? ()

Thread 6 (Thread 0x1603 of process 3484):
#0  0x00007fff69708ce6 in ?? ()
#1  0x00007fff697ca185 in ?? ()
--Type <RET> for more, q to quit, c to continue without paging--
#2  0x0000000000000000 in ?? ()

Thread 5 (Thread 0x1503 of process 3484):
#0  0x00007fff6970e5be in ?? ()
#1  0x00000001003571ef in caml_thread_tick ()
#2  0x00007fff697c9e65 in ?? ()
#3  0x0000000000000000 in ?? ()

Thread 4 (Thread 0x1103 of process 3484):
#0  0x00007fff69708ce6 in ?? ()
#1  0x00007fff697ca185 in ?? ()
#2  0x0000000000000000 in ?? ()

Thread 3 (Thread 0xf07 of process 3484):
#0  0x0000000100e78987 in ?? ()
#1  0x000000010033e235 in ocaml_avfilter_get_frame (_filter=4382129152) at avfilter_stubs.c:343
#2  0x000000010021e564 in camlFFmpeg__Avfilter__fun_729 ()
#3  0x0000000100029dcf in camlFfmpeg_decoder__flush_1265 ()
#4  0x0000000100029f56 in camlFfmpeg_decoder__flush_1265 ()
#5  0x00000001001b4479 in camlDecoder__fill_473 ()
#6  0x00000001000bbf04 in camlRequest_source__fun_1436 ()
#7  0x00000001001cba61 in camlTutils__mutexify_102 ()
#8  0x00000001001cba61 in camlTutils__mutexify_102 ()
#9  0x00000001000b9b1b in camlRequest_source__fun_1466 ()
--Type <RET> for more, q to quit, c to continue without paging--
#10 0x00000001001a82c2 in camlSource__fun_1297 ()
#11 0x0000000100095941 in camlOutput__fun_1230 ()
#12 0x00000001001a5dbd in camlClock__fun_1634 ()
#13 0x00000001002e41f8 in camlStdlib__list__fold_left_250 ()
#14 0x00000001001a3f31 in camlClock__fun_1585 ()
#15 0x00000001001a5a9d in camlClock__loop_930 ()
#16 0x00000001001a42a7 in camlClock__fun_1562 ()
#17 0x00000001001cbfb9 in camlTutils__fun_1467 ()
#18 0x00000001002d6e0b in camlThread__fun_286 ()
#19 0x0000000100382ed8 in caml_float_pcm_convert_s24le (_src=4382129152, _offset=4383227840, _dst=1099511628032) at audio_c.c:411
#20 0x0000000100357188 in caml_thread_start ()
#21 0x00007fff697c9e65 in ?? ()
#22 0x0000000000000000 in ?? ()

Thread 2 (Thread 0x1003 of process 3484):
#0  0x00007fff6970edee in ?? ()
#1  0x00007fff697ca348 in ?? ()
#2  0x0000000000000001 in ?? ()
#3  0x0000000000000000 in ?? ()
```

**Comment by toots** [comment]
Thanks for this! I've been able to identify an issue that might have been your problem. Would you mind trying with the latest `ocaml-ffmpeg`?

**Comment by Infernion** [comment]
Thanks for advice! Now I'm getting next error:

```
2019/12/15 21:56:27 [clock.main:3] Delegating synchronisation to CPU clock
liquidsoap(22553,0x7000048fb000) malloc: can't allocate region
*** mach_vm_map(size=117940271909490688) failed (error code=3)
liquidsoap(22553,0x7000048fb000) malloc: *** set a breakpoint in malloc_error_break to debug
Fatal error: out of memory
[Inferior 1 (process 22553) exited with code 02]
```

**Comment by toots** [comment]
Damn, sorry to hear. Any chance you could get a `gdb` trace again?

**Comment by Infernion** [comment]
I've run it with gdb already. The call to `thread apply all bt` have no effect

**Comment by toots** [comment]
Ok, thanks for trying it. What's your OS version?

**Comment by Infernion** [comment]
My OS is MACOS 10.15.2
null

**Comment by toots** [comment]
Damn, now I have to update my os for real.. 🙂

What version of `ffmpeg`? Installed via `homebrew`?

**Comment by toots** [comment]
BTW, if you have mac os you might want to try with `lldb` instead. There's an easy one-to-one mapping for simple commands: https://lldb.llvm.org/use/map.html

Chances are that `lldb` might be able to report more info.

**Comment by Infernion** [comment]
I have ffmpeg 4.2.1_2

I've tried to recreate opam env and reinstall everything.  Now I'm getting following error:

```
OCAMLOPT -c lang/builtins_dtools.ml
File "lang/builtins_dtools.ml", line 77, characters 46-71:
77 |           | Lang.String s -> make Lang.string (Dtools.Conf.string ~d:s)
                                                   ^^^^^^^^^^^^^^^^^^^^^^^^^
Error: This expression has type
         ?p:(Dtools.Conf.ut -> unit) ->
         ?l:Dtools.Conf.links ->
         ?comments:string list -> string -> string Dtools.Conf.t
       but an expression was expected of type
         ?p:(Dtools.Conf.ut -> unit) ->
         ?l:Dtools.Conf.links ->
         ?comments:string list ->
         string ->
         (< on_change : (string -> unit) -> 'weak2; ut : 'weak3;
            validate : (string -> bool) -> 'weak4; .. >
          as 'a)
       Type
         string Dtools.Conf.t =
           < alias : ?comments:string list ->
                     ?descr:string ->
                     (Dtools.Conf.ut -> unit) -> string Dtools.Conf.t;
             comments : string list; descr : string; get : string;
             get_d : string option; kind : string option;
             on_change : (string -> unit) -> unit;
             path : Dtools.Conf.path -> Dtools.Conf.ut;
             plug : Dtools.Conf.link -> Dtools.Conf.ut -> unit;
             routes : Dtools.Conf.ut -> Dtools.Conf.path list;
             set : string -> unit; set_d : string option -> unit;
             subs : Dtools.Conf.link list; ut : Dtools.Conf.ut >
       is not compatible with type
         < on_change : (string -> unit) -> 'weak2; ut : 'weak3;
           validate : (string -> bool) -> 'weak4; .. >
         as 'a
       The first object type has no method validate
make[2]: *** [lang/builtins_dtools.cmx] Error 2
make[1]: *** [all-auto-ocaml-prog] Error 2
make: *** [all-subdirs] Error 2
```

**Comment by smimram** [comment]
I am trying the youtube example to reproduce
```
output.url(url="rtmp://a.rtmp.youtube.com/live2/SECRET",encoder,s)
```
but this always results in
```
2019/12/18 08:44:54 [clock.main:2] Error when starting output(dot)url: Request_simple.Invalid_URI!
2019/12/18 08:44:54 [clock.main:3] Raised at file "sources/request_simple.ml", line 36, characters 56-73
2019/12/18 08:44:54 [clock.main:3] Called from file "source.ml", line 445, characters 8-20
2019/12/18 08:44:54 [clock.main:3] Called from file "operators/switch.ml", line 111, characters 10-21
2019/12/18 08:44:54 [clock.main:3] Called from file "list.ml", line 110, characters 12-15
2019/12/18 08:44:54 [clock.main:3] Called from file "source.ml", line 445, characters 8-20
2019/12/18 08:44:54 [clock.main:3] Called from file "outputs/output.ml", line 136, characters 6-22
2019/12/18 08:44:54 [clock.main:3] Called from file "source.ml", line 445, characters 8-20
2019/12/18 08:44:54 [clock.main:3] Called from file "clock.ml", line 348, characters 16-27
2019/12/18 08:44:54 [output(dot)url:1] Got ill-balanced activations (from output(dot)url)!
```
Any idea?

**Comment by toots** [comment]
> I have ffmpeg 4.2.1_2
> 
> I've tried to recreate opam env and reinstall everything. Now I'm getting following error:
> 
> ```
> OCAMLOPT -c lang/builtins_dtools.ml
> File "lang/builtins_dtools.ml", line 77, characters 46-71:
> 77 |           | Lang.String s -> make Lang.string (Dtools.Conf.string ~d:s)
>                                                    ^^^^^^^^^^^^^^^^^^^^^^^^^
> Error: This expression has type
>          ?p:(Dtools.Conf.ut -> unit) ->
>          ?l:Dtools.Conf.links ->
>          ?comments:string list -> string -> string Dtools.Conf.t
>        but an expression was expected of type
>          ?p:(Dtools.Conf.ut -> unit) ->
>          ?l:Dtools.Conf.links ->
>          ?comments:string list ->
>          string ->
>          (< on_change : (string -> unit) -> 'weak2; ut : 'weak3;
>             validate : (string -> bool) -> 'weak4; .. >
>           as 'a)
>        Type
>          string Dtools.Conf.t =
>            < alias : ?comments:string list ->
>                      ?descr:string ->
>                      (Dtools.Conf.ut -> unit) -> string Dtools.Conf.t;
>              comments : string list; descr : string; get : string;
>              get_d : string option; kind : string option;
>              on_change : (string -> unit) -> unit;
>              path : Dtools.Conf.path -> Dtools.Conf.ut;
>              plug : Dtools.Conf.link -> Dtools.Conf.ut -> unit;
>              routes : Dtools.Conf.ut -> Dtools.Conf.path list;
>              set : string -> unit; set_d : string option -> unit;
>              subs : Dtools.Conf.link list; ut : Dtools.Conf.ut >
>        is not compatible with type
>          < on_change : (string -> unit) -> 'weak2; ut : 'weak3;
>            validate : (string -> bool) -> 'weak4; .. >
>          as 'a
>        The first object type has no method validate
> make[2]: *** [lang/builtins_dtools.cmx] Error 2
> make[1]: *** [all-auto-ocaml-prog] Error 2
> make: *** [all-subdirs] Error 2
> ```

You need to pin to the latest `dtool` version.. Forgot to add the version dep on `liquidsoap.opam`, just did yesterday..

**Comment by toots** [comment]
> I am trying the youtube example to reproduce
> 
> ```
> s = single("test.ogv")
> encoder = %ffmpeg(video_codec="libx264",audio_codec="libmp3lame",channels=1,format="flv")
> output.url(url="rtmp://a.rtmp.youtube.com/live2/SECRET",encoder,s)
> ```
> 
> but this always results in
> 
> ```
> 2019/12/18 08:55:33 [single_25925:3] "test.ogv" is static, resolving once for all...
> 2019/12/18 08:55:33 [clock.main:2] Error when starting output(dot)url: Request_simple.Invalid_URI!
> 2019/12/18 08:55:33 [clock.main:3] Raised at file "sources/request_simple.ml", line 36, characters 56-73
> 2019/12/18 08:55:33 [clock.main:3] Called from file "source.ml", line 445, characters 8-20
> 2019/12/18 08:55:33 [clock.main:3] Called from file "outputs/output.ml", line 136, characters 6-22
> 2019/12/18 08:55:33 [clock.main:3] Called from file "source.ml", line 445, characters 8-20
> 2019/12/18 08:55:33 [clock.main:3] Called from file "clock.ml", line 348, characters 16-27
> ```
> 
> Any idea?
> 
> PS: yes, "test.ogv" is a valid file, I can play it locally (using Liquidsoap):
> 
> ```
> s = single("test.ogv")
> output.sdl(s)
> ```
> 
> is fine.

Yes, probably my fault. I was testing with a mono video. Try changing the `channels` param. Updating the code above..

**Comment by smimram** [comment]
In any case, we should have least have a more clear error message...

I have valgrined a bit, nothing shows up here.

**Comment by dhannyz** [comment]
@toots Something seems to have happened with output.url
I keep getting these errors
```
2020/01/13 09:33:20 [clock.main:2] Error when starting output output(dot)url: Failure("Not implemented.")!
2020/01/13 09:33:20 [clock.main:3] Raised by primitive operation at file "swscale/swscale.ml", line 75, characters 4-121
2020/01/13 09:33:20 [clock.main:3] Called from file "encoder/ffmpeg_encoder.ml", line 126, characters 10-110
2020/01/13 09:33:20 [clock.main:3] Called from file "tools/utils.ml", line 25, characters 38-43
2020/01/13 09:33:20 [clock.main:3] Called from file "encoder/ffmpeg_encoder.ml", line 113, characters 4-766
2020/01/13 09:33:20 [clock.main:3] Called from file "encoder/ffmpeg_encoder.ml", line 194, characters 14-48
2020/01/13 09:33:20 [clock.main:3] Called from file "encoder/ffmpeg_encoder.ml", line 201, characters 20-29
2020/01/13 09:33:20 [clock.main:3] Called from file "outputs/pipe_output.ml", line 66, characters 22-32
2020/01/13 09:33:20 [clock.main:3] Called from file "tools/start_stop.ml", line 104, characters 8-18
2020/01/13 09:33:20 [clock.main:3] Called from file "clock.ml", line 365, characters 20-38
```
